### ISSUE TITLE
redux-thunk "actions may not have undefined type property"

### ISSUE THREAD

**Issue by finaiized** [reduxjs/redux-mock-store#48]
Since upgrading to 1.1.0 or above (previously on 1.0.4), using redux-thunk, dispatching a thunk fails with "Actions may not have an undefined "type" property. Have you misspelled a constant?" (this was introduced in 2f7fa4400e).

For example, my action creator returns a thunk:

``` js
export function fetchSomething(smthg) {
  return dispatch => {
    return API.someCall(smthg).then(things => {
      dispatch(plainActionCreator(things));
      return things;
    });
  };
}
```

As well, I've configured the mock store to use `redux-thunk`.

My tests simply dispatch this action using a mock store:

``` js
describe('Blah', () => {
    it('should work', () => {
       return store.dispatch(fetchSomething('code'));
   });
});
```

**Comment by chrisblossom** [comment]
I am having this same problem. Please let me know if you find a solution.

**Comment by dmitry-zaets** [comment]
Can you please provide more code and may be error logs? Would be easier identify a problem.

**Comment by dmitry-zaets** [comment]
@finaiized Is there any specific reason why do you make `return things;` from action creator?
What is the code of `API.someCall(smthg)`?
Do you mock properly this call?
What is the code of `plainActionCreator`?

**Comment by finaiized** [comment]
@dmitry-zaets - Sorry, should have simplified the code. The only error message I get is "Actions may not have an undefined "type" property. Have you misspelled a constant?" after I call `mockStore.dispatch` with the above thunk. Basically, anytime I attempt to dispatch any thunk of form

``` js
export function thunkCreator() {
    return dispatch => //... do something
}
```

I run into this error.

All the other details (API calls, returns, etc) are just implementation details - they did work fine on 1.0.4 (and are mocked). For what it's worth, calling those functions in my app works fine and as expected. Also, all plain action creators work fine even after updating - this issue is limited to thunks only.

**Comment by dmitry-zaets** [comment]
@finaiized Problem is really in details.
They were working before because the mock store was not checking for empty actions. Empty actions are forbidden by Redux itself and all that we've added on that commit - is check from Redux.

Would be much easier to reproduce your issue if you will provide more data.

**Comment by aweary** [comment]
@finaiized can you show how you're creating your mock store? It should look like 

``` js
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

const middlewares = [thunk]; // add your middlewares like `redux-thunk`
const mockStore = configureStore(middlewares);
```

**Comment by finaiized** [comment]
Thanks for the help guys. Unsurprisingly, it was a problem on my end - as I was coming up with a repro case for this and couldn't, I looked closer at my existing tests. The problem wasn't actually in my thunk action creators, but in a couple of tests where I mocked them. I returned the _value_ of what I was expecting from my thunk action creators in the mock, failing to remember that I needed to return a _function_.

Eg.

``` js
export function thunkActionCreator() {
    return dispatch => Promise.resolve([]); // Do some real work here
}

// Wrong
// Returns a value
const stub = sinon.stub(actions, 'thunkActionCreator').returns(Promise.resolve([]));

// Right
// Returns a function
const stub = sinon.stub(actions, 'thunkActionCreator').returns(() => Promise.resolve([]));
```

Obviously, since my values have no `type`, it failed.

So thanks @dmitry-zaets, your commit fixed some of tests! :+1:
